### ISSUE TITLE
native support for complements

### ISSUE THREAD

**Issue by esuomi** [metosin/malli#109]
_As originally discussed on Clojurians Slack_

Currently creating a schema for matching non-empty strings requires the use of `:fn` for custom validation:

```clojure
[:and string? [:fn {:error/message "should not be empty"} not-empty]]
```
[_see live on malli.io_](https://malli.io/?value=%221%22&schema=%5B%3Aand%20string%3F%20%5B%3Afn%20%7B%3Aerror%2Fmessage%20%22should%20not%20be%20empty%22%7D%20not-empty%5D%5D)

This works, but is a bit ugly IMHO and also doesn't carry over to inferred schemas which might be an issue for some. Sidestepping a bit, with OpenAPI spec the recommended way to handle this is to set [string's `minLength` property to 1](https://swagger.io/docs/specification/data-models/data-types/#string).

It was suggested on Slack that malli could support [`not-empty`](http://clojuredocs.org/clojure.core/not-empty) out of the box which would be a good addition but as I believe that complements are useful in general there also could be something like a custom schema function for effectively complementing any part of schema, eg.
```clojure
[:and string? [:not empty?]]
```
Here I would prefer the word _not_ instead of _complement_ because for the latter as non-native English speaker and mostly self-taught I had to dig up a dictionary when I first encountered the word... 😛 I do recognize that while this sounds simple, it probably would have an impact on a lot of places in malli's internals as all bits from schema builders to humanize error message composition would need to be negation aware.

To summarize, for this issue malli could do one of these:

1. Add `not-empty` to malli's core
1. Implement `:not` (or `:complement`) schema function
1. Something totally different
1. Say the way it is now is good enough and do nothing at all :)

**Comment by ikitommi** [comment]
`:not` might be a good idea. We would better when/if that's implemented.

**Comment by ikitommi** [comment]
`:not´ most welcome.
